webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/volume/VolumeTracingDownsampling.scala

```diff
@@ -0,0 +1,209 @@
+  VersionedKeyValuePair
+}
+import com.scalableminds.webknossos.tracingstore.geometry.{Point3D => ProtoPoint3D}
+
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+import scala.reflect.ClassTag
+
+object VolumeTracingDownsampling {
+  def resolutionsForVolumeTracingByLayerName(dataSource: DataSourceLike,
+                                             fallbackLayerName: Option[String]): List[Point3D] = {
+    val fallbackLayer: Option[DataLayerLike] =
+      fallbackLayerName.flatMap(name => dataSource.dataLayers.find(_.name == name))
+    resolutionsForVolumeTracing(dataSource, fallbackLayer)
+  }
+
+  def resolutionsForVolumeTracing(dataSource: DataSourceLike, fallbackLayer: Option[DataLayerLike]): List[Point3D] = {
+    val fallBackLayerMags = fallbackLayer.map(_.resolutions)
+    fallBackLayerMags.getOrElse(dataSource.dataLayers.flatMap(_.resolutions).distinct)
+  }
+}
+
+trait VolumeTracingDownsampling
+    extends BucketKeys
+    with ProtoGeometryImplicits
+    with VolumeBucketCompression
+    with KeyValueStoreImplicits {
+
+  val tracingDataStore: TracingDataStore
+  val tracingStoreWkRpcClient: TracingStoreWkRpcClient
+  def saveBucket(dataLayer: VolumeTracingLayer,
+                 bucket: BucketPosition,
+                 data: Array[Byte],
+                 version: Long,
+                 toCache: Boolean = false): Fox[Unit]
+
+  private def fillMapWithInitialBucketsInplace(bucketDataMap: mutable.HashMap[BucketPosition, Array[Byte]],
+                                               tracingId: String,
+                                               dataLayer: VolumeTracingLayer): Unit = {
+    val data: List[VersionedKeyValuePair[Array[Byte]]] =
+      tracingDataStore.volumeData.getMultipleKeys(tracingId, Some(tracingId))
+    data.foreach { keyValuePair: VersionedKeyValuePair[Array[Byte]] =>
+      val bucketPosition = parseBucketKey(keyValuePair.key).map(_._2)
+      bucketPosition.foreach {
+        bucketDataMap(_) = decompressIfNeeded(keyValuePair.value,
+                                              expectedUncompressedBucketSizeFor(dataLayer),
+                                              s"bucket $bucketPosition during downsampling")
+      }
+    }
+  }
+
+  def downsampleWithLayer(tracingId: String, tracing: VolumeTracing, dataLayer: VolumeTracingLayer)(
+      implicit ec: ExecutionContext): Fox[Set[Point3D]] = {
+    val bucketVolume = 32 * 32 * 32
+    val originalMag = Point3D(1, 1, 1)
+    for {
+      requiredMags <- getRequiredMags(tracing)
+      elementClass = elementClassFromProto(tracing.elementClass)
+      bucketDataMap = new mutable.HashMap[BucketPosition, Array[Byte]]() {
+        override def default(key: BucketPosition): Array[Byte] = Array[Byte](0)
+      }
+      _ = fillMapWithInitialBucketsInplace(bucketDataMap, tracingId, dataLayer)
+      originalBucketPositions: List[BucketPosition] = bucketDataMap.keys.toList
+      updatedBuckets = new mutable.HashSet[BucketPosition]()
```

Comment:
Change to `List`

Reply:
@youri-k is this also resolved/a won't fix?

Reply:
I think it would be a nice to have, but it isn't necessary. If you want, I can make the change now.

Reply:
That's your call ;) Just wanted to check whether this can be resolved.